app/src/main/cpp/skyline/kernel/svc.cpp

```diff
@@ -21,241 +21,261 @@ namespace skyline::kernel::svc {
+            return;
+        }
 
-        if (!state.thread->switchPriority.Valid(priority)) {
-            state.ctx->registers.w0 = result::InvalidAddress;
+        if (!constant::HosPriority.Valid(priority)) {
+            state.ctx->gpr.w0 = result::InvalidPriority;
             state.logger->Warn("svcCreateThread: 'priority' invalid: {}", priority);
             return;
         }
 
-        auto thread{state.process->CreateThread(entryAddress, entryArgument, stackTop, priority)};
-        state.logger->Debug("svcCreateThread: Created thread with handle 0x{:X} (Entry Point: 0x{:X}, Argument: 0x{:X}, Stack Pointer: 0x{:X}, Priority: {}, TID: {})", thread->handle, entryAddress, entryArgument, stackTop, priority, thread->tid);
+        auto stack{state.process->GetMemoryObject(stackTop)};
+        if (!stack)
+            throw exception("svcCreateThread: Cannot find memory object in handle table for thread stack: 0x{:X}", stackTop);
+
+        auto thread{state.process->CreateThread(entry, entryArgument, stackTop, priority, idealCore)};
+        state.logger->Debug("svcCreateThread: Created thread with handle 0x{:X} (Entry Point: 0x{:X}, Argument: 0x{:X}, Stack Pointer: 0x{:X}, Priority: {}, ID: {})", thread->handle, entry, entryArgument, stackTop, priority, thread->id);
 
-        state.ctx->registers.w1 = thread->handle;
-        state.ctx->registers.w0 = Result{};
+        state.ctx->gpr.w1 = thread->handle;
+        state.ctx->gpr.w0 = Result{};
     }
 
-    void StartThread(DeviceState &state) {
-        auto handle{state.ctx->registers.w0};
+    void StartThread(const DeviceState &state) {
+        KHandle handle{state.ctx->gpr.w0};
         try {
             auto thread{state.process->GetHandle<type::KThread>(handle)};
-            state.logger->Debug("svcStartThread: Starting thread: 0x{:X}, PID: {}", handle, thread->tid);
+            state.logger->Debug("svcStartThread: Starting thread: 0x{:X}, PID: {}", handle, thread->id);
             thread->Start();
-            state.ctx->registers.w0 = Result{};
-        } catch (const std::exception &) {
+            state.ctx->gpr.w0 = Result{};
+        } catch (const std::out_of_range &) {
             state.logger->Warn("svcStartThread: 'handle' invalid: 0x{:X}", handle);
-            state.ctx->registers.w0 = result::InvalidHandle;
+            state.ctx->gpr.w0 = result::InvalidHandle;
         }
     }
 
-    void ExitThread(DeviceState &state) {
-        state.logger->Debug("svcExitThread: Exiting current thread: {}", state.thread->tid);
-        state.os->KillThread(state.thread->tid);
+    void ExitThread(const DeviceState &state) {
+        state.logger->Debug("svcExitThread: Exiting current thread: {}", state.thread->id);
+        std::longjmp(state.thread->originalCtx, true);
     }
 
-    void SleepThread(DeviceState &state) {
-        auto in{state.ctx->registers.x0};
+    void SleepThread(const DeviceState &state) {
+        u64 in{state.ctx->gpr.x0};
```

Comment:
I've changed all instances of a register value being taken as a `u64` into explicitly `u64` rather than `auto` as I think `auto` is just longer, it also aligns with using `u32` for all `WRegister` values because it calls `WRegister::operator u32()` for it.